### ISSUE TITLE
Fixed size batching

### ISSUE THREAD

**Issue by Matthieu-Tinycoaching** [bentoml/BentoML#1656]
Hello,

Would it be feasible to activate batching at a fixed size. As far as I know bentoML actually implement both non-batching and micro-bacthing which is adaptive one.

For certain architectures as AWS instances, the torchscript compiler is based on fixed batch size.

Thanks!

**Comment by bojiang** [comment]
We also considered providing a padding option for micro-batching. Then it would be possible to have fixed size batching.

Before an official solution, you can make full use of the freedom of the Python runtime, padding inputs by your own.

Here is an example:
```python
from itertools import cycle

FIXED_BATCH_SIZE = 5


@api(input=JsonInput(), mb_max_batch_size=FIXED_BATCH_SIZE)
def predict(self, json_list):
    length = len(json_list)
    if len(json_list) < FIXED_BATCH_SIZE:
        cycle_iter = cycle(json_list)
        json_list = [next(cycle_iter) for _ in range(FIXED_BATCH_SIZE)]. # cycle elements until FIXED_BATCH_SIZE
    results = self.artifacts.model(json_list)
    return results[:FIXED_BATCH_SIZE]
```

**Comment by Matthieu-Tinycoaching** [comment]
@bojiang thanks that seems a good workaround meanwhile the official solution !

**Comment by Matthieu-Tinycoaching** [comment]
@bojiang since I work with torch tensors, do you know which is the fastest between your solution using `cycle` Python runtime or with `torch.nn.ConstantPad1d`?

**Comment by Matthieu-Tinycoaching** [comment]
@bojiang if `len(json_list) < FIXED_BATCH_SIZE` shouldn't we return from the `predict` method `results[:len(json_list)]` instead of `results[:FIXED_BATCH_SIZE]` to take into account only the data of interest which were not padded?

In this case how to obtain from inference prediction of interest and not results from padded data?

**Comment by bojiang** [comment]
@Matthieu-Tinycoaching 
Oops yeah. It should be `results[:len(json_list)]`

**Comment by Matthieu-Tinycoaching** [comment]
@bojiang thanks.

Testing with this workaround gave me worse results. Wouldn't it be possible with the actual bentoML to have constantly batch of size `FIXED_BATCH_SIZE`, since with microbatching the batch size can vary between 1 and `FIXED_BATCH_SIZE`?

**Comment by bojiang** [comment]
> Testing with this workaround gave me worse results.

I didn't get it. "worse results" means bad performance or the api just doen't works?

**Comment by bojiang** [comment]
Would you please print actual batch sizes the api function got to the terminal?

**Comment by Matthieu-Tinycoaching** [comment]
Hi @bojiang 

I used the following predict API to pad tensors with zeros incomplete batch with `FIXED_BATCH_SIZE=6`:

```
    @api(input=JsonInput(), mb_max_batch_size=FIXED_BATCH_SIZE, batch=True)
    def predict(self, list_sentences: List[str]):
        tokenizer = self.artifacts.tokenizer
        encoded_input = tokenizer(list_sentences, padding='max_length', truncation=True, max_length=128, return_tensors='pt')

        # insert padding if we received a partial batch
        print(f"Size of batch = {len(list_sentences)}")
        padding = FIXED_BATCH_SIZE - len(list_sentences)
        if padding > 0:
            pad = torch.nn.ConstantPad1d((0, 0, 0, padding), value=0)
            input_batch = [pad(x) for x in (encoded_input['input_ids'], encoded_input['attention_mask'])]
        else:
            input_batch = [encoded_input['input_ids'], encoded_input['attention_mask']]
            
        with torch.no_grad():
            #Compute token embeddings
            model_output = self.artifacts.model(tensor=input_batch[0], tensor0=input_batch[1])
        return model_output[:len(list_sentences)].numpy()
```

And the actual batch size is always equal to `FIXED_BATCH_SIZE`, i.e. 6.

However, performances with `batch=False` (cf. `report_custom_bentoML128_250_spawn250_b1_worker1_inf1.xlarge.html`) is more stable (no fails) and has lower latency than with `batch=True` (cf. `report_custom_bentoML128_250_spawn250_b6_torchpad_worker1_inf1.xlarge.html`) in attached .zip file.

Thank you for your analysis!

[report_custom_bentoML128_250_spawn250.zip](https://github.com/bentoml/BentoML/files/6657186/report_custom_bentoML128_250_spawn250.zip)

**Comment by Asrst** [comment]
Hi, @bojiang,

I am looking forward to contribute this feature

With fixed size batching, batching service needs to be wait either till it reaches the requests equal to `mb_max_batch_size` or wait time equal to `mb_max_latency`. But I think it has its own limitations & might a optimal one.

can you please share your thoughts on it (or) guide how would you like to take this forward ?

**Comment by parano** [comment]
@Asrst the batching logic currently already does exactly this `wait either till it reaches the requests equal to mb_max_batch_size or wait time equal to mb_max_latency` - except that it estimate what's the latency for the request that arrives the earliest in the batch and make sure it can return within the `mb_max_latency`.

I think @Matthieu-Tinycoaching is asking for something a bit different, where the model runtime expect to receive the same batch size for every inference run. Always waiting for a fixed size before starting the inference run can cause timeout or unacceptable latency in some scenarios, e.g. under low traffic, a request may wait for very long time for other requests to come. One workaround I'd recommend is implementing a padding logic in front of the batching logic, exactly what you were trying to achieve that here: https://github.com/bentoml/BentoML/issues/1656#issuecomment-861643886  

As for the performance degrade that you pointed out, I think you may want to experiment with different batch sizes and explore more. BentoML by default will keep the historical run's batch size and execution time in memory, and dynamically use this data to adjust the batch size for better performance. Fixing the batch size means you lose that functionality which may lead to a worse performance than a simple non-batch solution.
